File: packages/template/src/runtime/prompt.ts
/**
 * System Prompt & Output Style Resolution
 *
 * Resolves the agent's system prompt across the ACP/config/file priority chain
 * and appends configured output styles. `withRuntimeContextPrompt` is exported
 * for the agent-config builder, which wraps the resolved prompt with runtime
 * workspace context.
 */

import { readFileSync, existsSync } from "node:fs";
import { resolve } from "node:path";
import { type AppConfig, type ACPSessionConfig } from "./config/config-loader.js";

/**
 * Resolve system prompt with priority chain:
 *   ACP session prompt > config.agent.systemPrompt > config.agent.systemPromptPath > inline fallback
 */
export function resolveSystemPrompt(
  config: AppConfig,
  sessionConfig: ACPSessionConfig | undefined,
  workspaceRoot: string
): string {
  // ACP session prompt takes highest priority
  if (sessionConfig?.systemPrompt) {
    return sessionConfig.systemPrompt;
  }

  if (config.agent.systemPrompt) {
    return withOutputStyle(config.agent.systemPrompt, config, workspaceRoot);
  }

  // Try loading from the configured prompt path.
  const promptPath = resolvePromptPath(config.agent.systemPromptPath, workspaceRoot);
  let basePrompt: string;
  if (existsSync(promptPath)) {
    const content = readFileSync(promptPath, "utf-8");
    // Strip the H1 title line (metadata, not prompt content)
    basePrompt = content.replace(/^# .*\r?\n/, "").trim();
  } else {
    // Inline fallback
    basePrompt = `You are ${config.agent.name} — an AI application agent.

## Workflow
1. Research — understand the task and check available tools
2. Plan — break down into steps using write_todos
3. Implement — execute each step
4. Verify — test and validate results

## Tool Priority (MANDATORY)
1. Platform MCP tools (query via platform_api) — ALWAYS check first
2. Built-in tools (http_request, platform_api, agent_variable, json_utils)
3. deepagents built-in tools (read_file, write_file, edit_file, execute, task)
4. Write custom code (last resort only)

## Rules
- When you need an external API key → create an agent variable
- When you need a tool → search platform plugins first
- Target agent prompts come from ACP — never hardcode them
- Save generated prompts via platform_api(operation: "save_prompt")
`;
  }

  // Append output style if configured
  return withOutputStyle(basePrompt, config, workspaceRoot);
}

/**
 * Resolve system prompt for CLI modes (REPL / one-shot).
 * Priority: explicit text > custom file > default prompt file > generic fallback.
 */
export function resolveCliSystemPrompt(options: {
  systemPrompt?: string;
  promptPath?: string;
  workspaceRoot?: string;
  config?: AppConfig;
}): string {
  if (options.systemPrompt) {
    return options.systemPrompt;
  }

  const workspaceRoot = options.workspaceRoot || process.cwd();
  if (options.promptPath) {
    const fullPath = resolvePromptPath(options.promptPath, workspaceRoot);
    if (existsSync(fullPath)) {
      return readFileSync(fullPath, "utf-8").replace(/^# .*\r?\n/, "").trim();
    }
  }

  if (options.config?.agent.systemPrompt) {
    return options.config.agent.systemPrompt;
  }

  const configuredPath = options.config?.agent.systemPromptPath;
  const defaultPath = resolvePromptPath(configuredPath || "prompts/developer-agent.system.md", workspaceRoot);
  if (existsSync(defaultPath)) {
    return readFileSync(defaultPath, "utf-8").replace(/^# .*\r?\n/, "").trim();
  }

  return "You are a helpful DeepAgent assistant. Be concise and action-oriented.";
}

/**
 * Load an output style file from prompts/styles/{name}.md.
 * Returns the style content (without frontmatter) or empty string if not found.
 */
export function resolveOutputStyle(styleName: string, workspaceRoot: string): string {
  const stylePath = resolve(workspaceRoot, "prompts/styles", `${styleName}.md`);
  if (!existsSync(stylePath)) {
    return "";
  }
  const content = readFileSync(stylePath, "utf-8");
  // Strip YAML frontmatter
  return content.replace(/^---\n[\s\S]*?\n---\n?/, "").trim();
}

function withOutputStyle(basePrompt: string, config: AppConfig, workspaceRoot: string): string {
  const style = resolveOutputStyle(config.agent.outputStyle, workspaceRoot);
  return style ? `${basePrompt}\n\n${style}` : basePrompt;
}

export function withRuntimeContextPrompt(basePrompt: string, workspaceRoot: string): string {
  return `${basePrompt}

## Runtime Context
- Effective workspace root: ${workspaceRoot}
- If the user asks for the current workspace directory, project root, cwd, runtime directory, or session location, use the \`runtime_info\` tool or answer from this Runtime Context.
- Do not infer the workspace by listing \`/\`, \`/Users\`, or parent directories.`;
}

function resolvePromptPath(path: string, workspaceRoot: string): string {
  if (path.startsWith("~/")) {
    return resolve(process.env.HOME || "", path.slice(2));
  }
  return path.startsWith("/") ? path : resolve(workspaceRoot, path);
}

File: packages/template-python/src/deepagents_template/runtime/prompt.py
"""System prompt assembly.

The Python template keeps the prompt logic in two layers:

* :func:`resolve_system_prompt` — picks CLI / REPL / ACP prompt sources.
* :func:`with_runtime_context_prompt` — appends runtime context (cwd, etc.).

The TS template additionally composes fragments from ``prompts/`` and
``prompts/styles/``; this file re-exports the resolver and provides a thin
``compose_prompt`` helper that callers can use to glue in user content.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from deepagents_template.runtime.config.config_schema import AppConfig


def compose_prompt(parts: list[str]) -> str:
    """Concatenate prompt fragments, separating them with a blank line."""
    return "\n\n".join(p.strip() for p in parts if p and p.strip())


def load_prompt_file(path: Path | str) -> str:
    """Read a prompt markdown file from disk."""
    return Path(path).read_text(encoding="utf-8")


def resolve_system_prompt(
    config: AppConfig,
    session_config: Any | None,
    workspace_root: Path | str,
) -> str:
    """Return the effective system prompt for a session.

    If the ACP session supplies a system prompt via metadata, it wins.
    Otherwise we fall back to the agent's configured style.
    """
    from deepagents_template.runtime.helpers import (
        resolve_cli_system_prompt,
    )

    if session_config is not None and getattr(session_config, "system_prompt", None):
        return str(session_config.system_prompt)
    return resolve_cli_system_prompt(config, workspace_root)


def with_runtime_context_prompt(prompt: str, workspace_root: Path | str) -> str:
    """Append runtime context (workspace root, env hints) to *prompt*."""
    from deepagents_template.runtime.helpers import with_runtime_context_prompt as _with

    return _with(prompt, workspace_root)
